sdk/src/network/mod.rs

```diff
@@ -116,6 +121,196 @@ pub trait Network: Send + Sync + Clone + 'static {
     }
 }
 
+pub struct NetworkMultiplexer {
+    to_receiving_streams: ActiveStreams,
+    unclaimed_receiving_streams: Arc<DashMap<StreamKey, MultiplexedReceiver>>,
+    tx_to_networking_layer: MultiplexedSender,
+}
+
+type ActiveStreams = Arc<DashMap<StreamKey, tokio::sync::mpsc::UnboundedSender<ProtocolMessage>>>;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Default, Serialize, Deserialize)]
+pub struct StreamKey {
+    pub job_id: u8,
+    pub task_hash: [u8; 32],
+    pub round_id: u16,
+}
+
+pub struct MultiplexedReceiver {
+    inner: tokio::sync::mpsc::UnboundedReceiver<ProtocolMessage>,
+    stream_id: StreamKey,
+    // For post-drop removal purposes
+    active_streams: ActiveStreams,
+}
+
+#[derive(Clone)]
+pub struct MultiplexedSender {
+    inner: tokio::sync::mpsc::UnboundedSender<(StreamKey, ProtocolMessage)>,
+    pub(crate) stream_id: StreamKey,
+}
+
+impl MultiplexedSender {
+    pub fn send(&self, message: ProtocolMessage) -> Result<(), Error> {
+        self.inner
+            .send((self.stream_id, message))
+            .map_err(|err| Error::Other(err.to_string()))
+    }
+}
+
+impl Stream for MultiplexedReceiver {
+    type Item = ProtocolMessage;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Pin::new(&mut self.get_mut().inner).poll_recv(cx)
+    }
+}
+
+impl Deref for MultiplexedReceiver {
+    type Target = tokio::sync::mpsc::UnboundedReceiver<ProtocolMessage>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl Drop for MultiplexedReceiver {
+    fn drop(&mut self) {
+        let _ = self.active_streams.remove(&self.stream_id);
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct MultiplexedMessage {
+    payload: Vec<u8>,
+    stream_id: StreamKey,
+}
+
+impl NetworkMultiplexer {
```

Comment:
Could NetworkMultiplexer itself impl Network too? this would make it seamless for the end user to switch between the normal implementation and the multiplixed one.

Reply:
We wouldn't want to expose the normal implementation (i.e., N: Network = GossipHandle) to the user. In fact, this is a level above our `N` implementation, so it could lead to mixups. So, if we have two layers of the networking stack competing to use `N` (i.e., the multiplexed layer and the "normal" layer, `N`), it wouldn't work. So, what we could do now that this test is working is:

Wrap each GossipHandle inside the `NetworkMultiplexer`. Remove any Clone implementations on the GossipHandle to enforce a single GossipHandle in memory.

`NetworkMultiplexer` could be made to implement `Clone`. Each time we start a job, we can pass the multiplexer to the user. They can then get a tx/rx handle by calling `NetworkMultiplexer::multiplex`. Note: we can perhaps abstract some of this away by passing a simpler handle to the user that removes the requirement of knowing what multiplexing is, and instead, just have a subset of `NetworkMultiplexer` called `JobNetwork`, and let them obtain a tx/rx handle via `JobNetwork::new_round() -> NetworkRound`. In this case, `NetworkRound` could implement `Network`.